### ISSUE TITLE
Are multilingual tests possible?

### ISSUE THREAD

**Issue by fasyoki** [marcusgreen/moodle-qtype_gapfill#87]
Hello,

I am trying to use the {mlang} tags for our multilingual Moodle page.
Is there a way to make that work as yet it seems that the quiz with Gapfill question type cannot identify these tags properly.
Sorry if my question is not specific or detailed enough. I am not a proper developer obviously.

![image](https://user-images.githubusercontent.com/44682506/173306695-66ee6ac1-c029-4d3d-8b39-2bdd5543ca28.png)

If you have any suggestions, I would be glad to hear them.
Thanks,
Gabor

**Comment by marcusgreen** [comment]
You ask a good question. My expectation is that the question type will work with both the mlang and the mlang2 filters
(https://moodle.org/plugins/filter_multilang2).

You can see me talking about it at the 2019 Japan MoodleMoot 
https://youtu.be/7sZh33dppIs?t=2501

Can you confirm the version of Gapfill you are using, the version of Moodle and if will process {mlang} strings in other places on your course (such as a label).

If you cannot find those version numbers, please export an example of one of your questions in xml format and attach it to a comment here.
Best wishes
Marcus

**Comment by fasyoki** [comment]
Hello Marcus,

Many thanks for being this quick with your response. 
I assume I am using 3.8.1 build (definitely not the latest), hopefully I am checking it at the right place:

![image](https://user-images.githubusercontent.com/44682506/173312759-2217f3f2-9ef1-43f8-b2a6-881dc6d2495f.png)

While the Gapfill has this version number (the latest I presume):

![image](https://user-images.githubusercontent.com/44682506/173313078-37463d79-6452-449e-bbec-760dad998e55.png)

I started with the default 'drag and drop to text' quiz type, but that did not allow mlang2 filters (or I just failed with it somehow...), so I switched to the 'select missing words' type, which worked fine with the multilingual behaviour of the site, but then I realized that I just could not erase any of the empty choices from the list and so I always ended up with partly completed answers. While looking for the solution, it was recommended somewhere that Gapfill could fill this gap (sorry, I could not help myself here..) So I crossed my fingers and gave it a try.

Can it happen that if I choose an earlier version of Gapfill that matches the Moodle build more, I can have a better result?

Best wishes,
Gabor

**Comment by fasyoki** [comment]
Thanks Marcus.

I have just tried two things:

I removed the newer version and installed the one that matches my build. The other thing was that I removed the { } types from the choosable delimit charachters (in the general settings). Yet no success..

![image](https://user-images.githubusercontent.com/44682506/173330763-f2289c1c-0d4d-4966-a91e-04a1299b729a.png)

![image](https://user-images.githubusercontent.com/44682506/173331351-3363dbc4-9d92-446e-9f31-6970bbc93344.png)

The graphical realisation of the plugin is very nice though!
I would love to switch to your test type, but this tricky mlang2 just does not allow it. 
If you still happen to find a solution, please let me know.

Many thanks,
Best wishes,
Gabor

**Comment by marcusgreen** [comment]
Here is an example of how you can use the core mlang filter with gapfill
<p>This illustrates the use of the core mlang filter (not Mlang2)</p>
<p>The english word for</p>
&lt;span lang=&quot;fr&quot; class=&quot;multilang&quot;&gt;un&lt;/span&gt;<br/>
&lt;span lang=&quot;es&quot; class=&quot;multilang&quot;&gt;uno&lt;/span&gt;<br/>
&lt;span lang=&quot;de&quot; class=&quot;multilang&quot;&gt;ein&lt;/span&gt;<br/>
is [one].
<p>If the current language selected is neither french,spanish or german the first item will be selected and un will be shown.</p>

**Comment by fasyoki** [comment]
Thanks, Marcus.
I tested it but it still does not work for me. It could happen that mlang filter does not work for me in general though.
I would need some more test, yet I could only check if the filter is set on:

![image](https://user-images.githubusercontent.com/44682506/173938686-22251547-6d8c-450d-9e07-842716c5baaf.png)

**Comment by marcusgreen** [comment]
It looks like you have both filters on, I suggest trying one at a time starting with core mlang. The thing that confused me was that if you don't have one of the languages selected from the dropdown it will show whatever occurs first in the text.  
So in my example with the default english language selected it just showed the first option un.
I can give you teacher access to a course if it would help. Let me know.

**Comment by marcusgreen** [comment]
I am closing this as I was not able to reproduce it. Please come back to me if it is still an issue and I will investigate further.

**Comment by fasyoki** [comment]
Dear Marcus,

I think the problem might have been that I used the 'codes' in a wrong way.
Even in a multilingual setting the [1], [2] etc. tags should appear only once as part of the Question text.
So a proper way is the following:

{mlang en}Question part 1 in English{mlang}
{mlang pt}Question part 1 in Portuguese{mlang}
[1]
{mlang en} question part 2 in English{mlang}
{mlang pt} question part 2 in Portuguese{mlang}
[2]
{mlang en} question part 3 in English{mlang}
{mlang pt} question part 3 in Portuguese{mlang}

The wrong way is the following for duplicating the tags for no reason:

{mlang en}Question part 1 in English [1] question part 2 in English [2] question part 3 in English.{mlang}
{mlang pt}Question part 1 in Portuguese [1] question part 2 in Portuguese [2] question part 3 in Portuguese.{mlang}

So sorry for the confusion and for your time!
My wording was completely wrong and illogical, that's why I had problems with the plugin, I believe.

Best wishes,
Gabor

**Comment by fasyoki** [comment]
Or to be more precise, I don't know if the right wording would help to avoid the problems occured with Gapfill.
As it solved my problem with the default question type of Moodle ('Select the missing' word or similar), I used that one finally and did not go back to Gapfill.

**Comment by marcusgreen** [comment]
Many thanks for your feedback, it was most appreciated. There have been many times when I have tried to make multi lang filters work and I found it didn't work as I expected. I found the 3rd party mlang2 was more predictable. The core Cloze question type is very good for fill in the blanks questions, but the syntax is quiet complex.However there are 3rd party editors that make it easier to use.  

I will re-visit the ease of use and documentation of language filters in Gapfill and Wordselect at some point in the future.
Best wishes
Marcus
